hypothesis-python/src/hypothesis/extra/array_api.py

```diff
@@ -243,6 +248,26 @@ def check_valid_minmax(prefix, val, info_obj):
                 check_valid_interval(min_value, max_value, "min_value", "max_value")
             kw["max_value"] = max_value
 
+        # By default, floats() will generate subnormals if they are in the
+        # inferred values range. If we have detected that xp flushes to zero for
+        # the passed dtype, we ensure floats() will not generate subnormals
+        # unless allow_subnormal=True is passed.
+        if allow_subnormal is not None:
+            kw["allow_subnormal"] = allow_subnormal
+        elif widths_ftz[finfo.bits]:
+            # floats() does not accept allow_subnormal=False when subnormals are
+            # out of range, so we only pass False when in range.
+            if min_value is None and max_value is None:
+                kw["allow_subnormal"] = False
+            else:
+                inferred_min = min_value or 0
+                inferred_max = max_value or 0
+                if (
+                    inferred_min > -finfo.smallest_normal
+                    or inferred_max < finfo.smallest_normal
```

Comment:
I think you've negated the wrong one here: we can't `allow_subnormal` if our min is larger than *positive* smallest_normal.  I'd also be inclined to inline the vars, like `(min_value or 0) > finfo.smallest_normal`.

Reply:
min/max confusion strikes again... I _think_ that I'm correct here, as this code block is purposely disabling `allow_subnormal` if they would otherwise be generated (when FTZ), i.e. when `-finfo.smallest_normal < min_value` or `max_value < finfo.smallest_normal`.

Note that `floats()` should not allow you to pass `allow_subnormals=False` when subnormals are not in the inferred values range anyway, like is the case with `allow_nan/infinity`. That's actually not the case right now, will fix/test.

Agree on inline vars.

Also noticed I don't account for exclude min/max, will need to add and test that behaviour.

Reply:
Nice catch on the exclusive endpoints!  I think this is probably covered when we adjust the min_value and max_value, though maybe we should conditionally use next_down_normal etc.?

`floats(1, 2, allow_infinity=False).example()` actually works just fine, it's only if you pass `=True` that it's inconsistent (and we don't error on redundancy).  So this might make your logic a bit simpler 🙂

Reply:
Oh my, for some reason I misunderstood the `allow_*` behaviour. Thanks for telling me. Wasted so much time on this :upside_down_face: